### ISSUE TITLE
Orestar transaction type/sub-type mapping

### ISSUE THREAD

**Issue by andrewbiang888** [hackoregon/openelections#970]
# Description
The goal of this issue is to define how the transaction type and sub-types map to the data Orestar expects to receive. 

## TLDR; Here are the questions:
1. Does the OAE type "Contribution" map to the Orestar type "C":
- [x] yes
- [ ] no. It is ___
2. Does the OAE type "Other Receipt" map to the Orestar type "OR"
- [x] yes
- [ ] no. It is ___
3. What Orestar sub-type codes do the following OAE sub-types map to?

| OAE        | Orestar           |
| ------------- |:-------------|
| Cash Contribution      | CA |
| In-Kind Contribution     |  IK    |
| In-Kind Contribution - Paid Supervision of Volunteers* | ? |
| In-Kind Forgiven Accounts Payable | IKA |
| In-Kind /Forgiven Personal Expenditure | IKP |
| Items Sold at Fair Market Value | FM |
| Lost or Returned Check | LC |
| Miscellaneous Other Receipt | OM |
| Refunds and Rebates | RF |

_*not sure where this came from but can't seem to find it in frontend anymore_
_for Orestar sub-type options, see Orestar -> Sub-types xml below_

# More Information
## Our App
Currently, as defined in[ this google doc](https://docs.google.com/document/d/1Fkx_2fqXICQ50ldenxymAz7GqMmlCGeOUiU8YXNHnJY/edit), the OAE app currently asks for two types initially and shows 2 sets of sub-types depending on what is selected.

### Types
The two types are 1) Contribution, and 2) Other Receipt. 

![Screen Shot 2019-10-17 at 1 46 06 PM](https://user-images.githubusercontent.com/20327807/67046182-a13f4a00-f0e4-11e9-8010-d88fc7a90b8a.png)
![Screen Shot 2019-10-17 at 1 46 30 PM](https://user-images.githubusercontent.com/20327807/67046191-a43a3a80-f0e4-11e9-95dd-d429aaeaee12.png)

### Sub-Types
There are a few more sub-types than types that we are tracking. Here is the description from the google doc linked above:
> If Contribution was selected, drop down says: Cash Contribution, In-Kind Contribution, In-Kind Contribution - Paid Supervision of Volunteers, In-Kind Forgiven Accounts Payable, In-Kind /Forgiven Personal Expenditure. If Other Receipt was selected, drop down says: Items Sold at Fair Market Value, Lost or Returned Check, Miscellaneous Other Receipt, Refunds and Rebates.

We are indeed asking this questions correctly on the front end, although "In-Kind Contribution - Paid Supervision of Volunteers" seems to be MIA when "Contribution" is selected for the type.
![Screen Shot 2019-10-17 at 2 03 28 PM](https://user-images.githubusercontent.com/20327807/67047408-1b70ce00-f0e7-11e9-975a-11f174ffd221.png)
![Screen Shot 2019-10-17 at 2 04 05 PM](https://user-images.githubusercontent.com/20327807/67047413-1e6bbe80-f0e7-11e9-9e1c-09dbec58b703.png)

## Orestar
Orestar's types are a little more robust that the types the OAE app asks for. 

### Types
Here is the XML for transaction types:
```xml
<xs:element name="type">
 <xs:annotation>
   <xs:documentation>The general category of the transaction.</xs:documentation>
  </xs:annotation>
  <xs:simpleType>
   <xs:restriction base="xs:string">
    <xs:enumeration value="C"/>
    <xs:enumeration value="E"/>
    <xs:enumeration value="OR"/>
    <xs:enumeration value="OD"/>
    <xs:enumeration value="OA"/>
    <xs:enumeration value="O"/>
   </xs:restriction>
  </xs:simpleType>
</xs:element>
```
In other words, Orestar expects the type to be one of the 6: C, E, OR, OD, OA, or O.

We need to confirm that:
- [ ] the OAE types map to the Orestar types like this: "Contribution" -> "C", and "Other Receipt" -> "OR". 
- [ ] the sub-types are actually associated with one of those two.

### Sub-types
Here is the XML for sub-types:
```xml
<xs:element name="sub-type">
  <xs:annotation>
   <xs:documentation>The specific category of the transaction; the subtype must be related to
the type.</xs:documentation>
  </xs:annotation>
  <xs:simpleType>
   <xs:restriction base="xs:string">
   <xs:enumeration value="CA"/>
   <xs:enumeration value="IK"/>
   <xs:enumeration value="IKP"/>
   <xs:enumeration value="IKA"/>
   <xs:enumeration value="NLR"/>
   <xs:enumeration value="PL"/>
   <xs:enumeration value="PI"/>
   <xs:enumeration value="PC"/>
   <xs:enumeration value="ELR"/>
   <xs:enumeration value="FM"/>
   <xs:enumeration value="IN"/>
   <xs:enumeration value="OM"/>
   <xs:enumeration value="RF"/>
   <xs:enumeration value="LC"/>
   <xs:enumeration value="OR"/>
   <xs:enumeration value="AE"/>
   <xs:enumeration value="PE"/>
   <xs:enumeration value="CE"/>
   <xs:enumeration value="AP"/>
   <xs:enumeration value="NLP"/>
   <xs:enumeration value="ELP"/>
   <xs:enumeration value="NP"/>
   <xs:enumeration value="RT"/>
   <xs:enumeration value="UIP"/>
   <xs:enumeration value="UCP"/>
   <xs:enumeration value="ULP"/>
   <xs:enumeration value="NLF"/>
   <xs:enumeration value="APR"/>
   <xs:enumeration value="OMD"/>
   <xs:enumeration value="CBA"/>
   <xs:enumeration value="PEA"/>
  </xs:restriction>
 </xs:simpleType>
</xs:element>
```
As you can see there are many more options here. 31 to be exact. These sub-types need to be mapped to the OAE sub-types described above.

What Orestar is expecting may be slightly different that the way the OAE app collects this data, so that will need to be confirmed as well.

**Comment by danmelton** [comment]
@susanmottet @amysantee @gdoucette we could use a hand!

**Comment by andrewbiang888** [comment]
Also, it is possible that "Other Receipt" just refers to "O", or other. The enum we defined in the api maps it to other at least:

![Screen Shot 2019-10-17 at 2 59 34 PM](https://user-images.githubusercontent.com/20327807/67050998-c0db7000-f0ee-11e9-83b1-7f1158adb677.png)

**Comment by amysantee** [comment]
This is outside my knowledge of ORESTAR which is minimal. Sorry!

**Comment by swoolley** [comment]
https://sos.oregon.gov/elections/Documents/orestarTransFiling.pdf might illuminate some of this on pages 31 and 32.

I made my most educated guess on how the mapping works out, and this is it.  Below is the xml with abbreviations inserted that I used to do process of elimination.

```
Transaction TypeAvailable Transaction SubTypes

Contribution (type=C)
  Cash Contribution (sub-type=CA)
  In-Kind Contribution (sub-type=IK)
  In-Kind/Forgiven Account Payable (sub-type=IKA)
  In-Kind/Forgiven Personal Expenditure (sub-type=IKP)
  Loan Received (Non-Exempt) (sub-type=NLR)
Expenditure (type=E)
  Account Payable (sub-type=AP)
  Cash Expenditure (sub-type=CE)
  Loan Payment (Non-Exempt) (sub-type=NLP)
  Personal Expenditure for Reimbursement (sub-type=PE)
Other (type=O)
  Account Payable Rescinded (sub-type=APR)
  Cash Balance Adjustment (sub-type=CBA)
  Loan Forgiven(Non-Exempt) (sub-type=NLF)
  Personal Expenditure Balance Adjustment (sub-type=PEA)
Other Account Receivable (type=OA)
  Miscellaneous Account Receivable (sub-type=OR)
Other Disbursement (type=OD)
  Loan Payment (Exempt) (sub-type=ELP)
  Miscellaneous Other Disbursement (sub-type=OMD)
  Nonpartisan Activity (sub-type=NP)
  Return or Refund of Contribution (sub-type=RT)
Other Receipt (type=OR)
  Interest/Investment Income (sub-type=IN)
  Items Sold at Fair Market Value (sub-type=FM)
  Loan Received (Exempt) (sub-type=ELR)
  Lost or Returned Check (sub-type=LC)
  Miscellaneous Other Receipt (sub-type=OM)
  Refunds and Rebates (sub-type=RF)
```

I don't know what AE means.  I also don't know why OMD would appear out of order.  I figured out where the main categories were buy finding the most obvious ones and then noticed nearby ones seemed to be in the nearest categories.  There are some pledge entries that they repealed from orestar that are no longer documented but still supported in the old schema.

```
C
   <xs:enumeration value="CA"/>C
   <xs:enumeration value="IK"/>C
   <xs:enumeration value="IKP"/>C
   <xs:enumeration value="IKA"/>C
   <xs:enumeration value="NLR"/>C
(Pledges were deleted from orestar)
   <xs:enumeration value="PL"/>
   <xs:enumeration value="PI"/>
   <xs:enumeration value="PC"/>
OR
   <xs:enumeration value="ELR"/>OR
   <xs:enumeration value="FM"/>OR
   <xs:enumeration value="IN"/>OR
   <xs:enumeration value="OM"/>OR
   <xs:enumeration value="RF"/>OR
   <xs:enumeration value="LC"/>OR
OA
   <xs:enumeration value="OR"/>OA
   <xs:enumeration value="AE”/>???????
E
   <xs:enumeration value="PE"/>E
   <xs:enumeration value="CE"/>E
   <xs:enumeration value="AP"/>E
   <xs:enumeration value="NLP"/>E
OD
   <xs:enumeration value="ELP"/>OD
   <xs:enumeration value="NP"/>OD
   <xs:enumeration value="RT"/>OD
(Pledges were deleted from orestar)
   <xs:enumeration value="UIP"/>
   <xs:enumeration value="UCP"/>
   <xs:enumeration value="ULP"/>
O
   <xs:enumeration value="NLF"/>O
   <xs:enumeration value="APR"/>O
   <xs:enumeration value="OMD"/>OD??
   <xs:enumeration value="CBA"/>O
   <xs:enumeration value="PEA"/>O
```

**Comment by swoolley** [comment]
If we're unsure of the mappings, I can actually do an xml upload in one of my accounts and delete the imported transactions, but before I do a delete I can see inside orestar what they mapped to.  In case there's any questions, that's how I'd answer them.
